In [3]:
import numpy as np

# Define scalar and vector size
a = 3
N = 10

# Create vectors
x = np.full(N, 0.1)  # x = [0.1, 0.1, ..., 0.1]
y = np.full(N, 7.1)  # y = [7.1, 7.1, ..., 7.1]

# Compute d = a * x + y
d = a * x + y

# Print result
print("Results for d = a * x + y with N = 10:")
for i in range(N):
    print(f"d[{i}] = {d[i]}")


Results for d = a * x + y with N = 10:
d[0] = 7.3999999999999995
d[1] = 7.3999999999999995
d[2] = 7.3999999999999995
d[3] = 7.3999999999999995
d[4] = 7.3999999999999995
d[5] = 7.3999999999999995
d[6] = 7.3999999999999995
d[7] = 7.3999999999999995
d[8] = 7.3999999999999995
d[9] = 7.3999999999999995


for N= 10e6

In [ ]:
a = 3
N = int(1e6)
x=np.full(N,0.1)
y=np.full(N,7.4)
d=a*x+y
print("Results for d = a * x + y with N = 10E6:")
for i in range(N):
    print(f"d[{i}] = {d[i]}")

The above code worked but the problem came with long run time, it started printing all ouctomes of d for 10e6 times. To solve this problem, here is the modified version of code. 

In [5]:
import time

N = int(1e6)
a = 3

x = np.full(N, 0.1, dtype=np.float32)
y = np.full(N, 7.1, dtype=np.float32)

# To measure computation time and get more faster results, added this function
start = time.time()
d = a * x + y
end = time.time()

# To verify the correctness of code that our computed result comes equal 7.4
correct = np.allclose(d, 7.4)

print(f"N = {N:,}")
print(f"All values correct: {correct}")
print(f"Computation time: {end - start:.4f} seconds")


N = 1,000,000
All values correct: True
Computation time: 0.0041 seconds


The above modifications reduced the computation time surprizingly. On a cluster of 1 NVIDIA GPU, it took only 0.0041 seconds, where before it would have taken around few minutes. 

In [6]:
import time

N = int(1e8)
a = 3

x = np.full(N, 0.1, dtype=np.float32)
y = np.full(N, 7.1, dtype=np.float32)

# To measure computation time and get more faster results, added this function
start = time.time()
d = a * x + y
end = time.time()

# To verify the correctness of code that our computed result comes equal 7.4
correct = np.allclose(d, 7.4)

print(f"N = {N:,}")
print(f"All values correct: {correct}")
print(f"Computation time: {end - start:.4f} seconds")


N = 100,000,000
All values correct: True
Computation time: 0.3517 seconds


For 10E8 dimension, it took only 0.3517 seconds. 

Part 3 of this task. Matrix multiplication C=AB for N= 10, 100 and 100000. A=3, B= 7.1

In [ ]:
import numpy as np

# Define matrix size and values
N = 10000
A = np.full((N, N), 3.0)
B = np.full((N, N), 7.1)

# Compute matrix product
C = np.matmul(A, B)

# Print the result
print("Matrix C = A × B:")
print(C)


Matrix C = A × B:
[[213000. 213000. 213000. ... 213000. 213000. 213000.]
 [213000. 213000. 213000. ... 213000. 213000. 213000.]
 [213000. 213000. 213000. ... 213000. 213000. 213000.]
 ...
 [213000. 213000. 213000. ... 213000. 213000. 213000.]
 [213000. 213000. 213000. ... 213000. 213000. 213000.]
 [213000. 213000. 213000. ... 213000. 213000. 213000.]]
Sample values from A and B:
A[0,0] = 3.0
B[0,0] = 7.1


In [1]:
import numpy as np
import time

def run_matrix_test(N):
    print(f"Testing matrix multiplication for N = {N}...")

    # Initialize matrices
    A = np.full((N, N), 3.0, dtype=np.float32)
    B = np.full((N, N), 7.1, dtype=np.float32)

    # Time the multiplication
    start = time.time()
    C = np.matmul(A, B)
    end = time.time()

    # Expected value for each element in C
    expected = 21.3 * N
    correct = np.allclose(C, expected, atol=1e-5)

    # Output
    print(f"Expected each element: {expected}")
    print(f"All elements correct: {correct}")
    print(f"Computation time: {end - start:.4f} seconds\n")

# Test for different N
for N in [10, 100, 10000]:
    run_matrix_test(N)


Testing matrix multiplication for N = 10...
Expected each element: 213.0
All elements correct: True
Computation time: 0.0004 seconds

Testing matrix multiplication for N = 100...
Expected each element: 2130.0
All elements correct: True
Computation time: 0.0001 seconds

Testing matrix multiplication for N = 10000...
Expected each element: 213000.0
All elements correct: True
Computation time: 1.1080 seconds

